client/components/hoc/protect-form/index.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import debugModule from 'debug';
+import page from 'page';
+import i18n from 'i18n-calypso';
+import { without } from 'lodash';
+
+/**
+ * Module variables
+ */
+const debug = debugModule( 'calypso:protect-form' );
+const confirmText = i18n.translate( 'You have unsaved changes. Are you sure you want to leave this page?' );
+const beforeUnloadText = i18n.translate( 'You have unsaved changes.' );
+let formsChanged = [];
+
+export const protectForm = WrappedComponent => {
+	class ProtectedFormComponent extends Component {
+		componentDidMount() {
+			window.addEventListener( 'beforeunload', this.warnIfChanged );
+		}
+
+		componentWillUnmount() {
+			window.removeEventListener( 'beforeunload', this.warnIfChanged );
+			formsChanged = without( formsChanged, this );
+		}
+
+		warnIfChanged = event => {
+			if ( ! formsChanged.length ) {
+				return;
+			}
+			debug( 'unsaved form changes detected' );
+			( event || window.event ).returnValue = beforeUnloadText;
+			return beforeUnloadText;
+		};
+
+		markChanged = () => {
+			if ( -1 === formsChanged.indexOf( this ) ) {
+				formsChanged.push( this );
+			}
+		};
+
+		markSaved = () => {
+			const index = formsChanged.indexOf( this );
+			if ( -1 === index ) {
+				return;
+			}
+			formsChanged.splice( index, 1 );
+		}
+
+		render() {
+			return (
+				<WrappedComponent
+					markChanged={ this.markChanged }
+					markSaved={ this.markSaved }
+					{ ...this.props }
+				/>
+			);
+		}
+	}
+
+	return ProtectedFormComponent;
+};
+
+export const checkFormHandler = ( context, next ) => {
```

Comment:
I'm not familiar with how we use `page.js`, It would be possible if all the navigation links and redirection use the `ROUTE_SET` action to trigger the navigation. I think it's not the case currently, it's the opposite, page.js triggers `ROUTE_SET` so It's too late for me. to avoid the navigation. (BTW, we need someone really familiar with our `page.js` setup here)